### ISSUE TITLE
Skill Balance sometimes horribly misbalances teams

### ISSUE THREAD

**Issue by NotJustin** [NotJustin/SkillAutoBalance#8]
See: https://github.com/NotJustin/SkillAutoBalance/issues/2#issuecomment-636507391
Need to debug. It is definitely related to this section:

```
void SortCloseSums(int outliers)
{
	int client, team;
	int i = 0;
	int size = GetClientCountNoBots() / 2 - outliers;
	float tSum = 0.0;
	float ctSum = 0.0;
	int tCount = 0;
	int ctCount = 0;
	while(tCount < size && ctCount < size)
	{
		client = g_iClients[i];
		if (client != 0 && IsClientInGame(client) && !IsFakeClient(client) && (team = GetClientTeam(client)) != TEAM_SPEC && team != UNASSIGNED && !g_iOutlierClients[client])
		{
			if (tSum < ctSum)
			{
				tSum += g_iClientScore[client];
				++tCount;
				if (g_iClientTeam[client] == TEAM_CT)
				{
					SwapPlayer(client, TEAM_T, "Client Skill Balance");
				}
			}
			else
			{
				ctSum += g_iClientScore[client];
				++ctCount;
				if (g_iClientTeam[client] == TEAM_T)
				{
					SwapPlayer(client, TEAM_CT, "Client Skill Balance");
				}
			}
		}
		++i;
	}
	while(i < sizeof(g_iClients))
	{
		client = g_iClients[i];
		if (client != 0 && IsClientInGame(client) && !IsFakeClient(client) && (team = GetClientTeam(client)) != TEAM_SPEC && team != UNASSIGNED)
		{
			if (tCount < size)
			{
				if (g_iClientTeam[client] == TEAM_CT)
				{
					SwapPlayer(client, TEAM_T, "Client Skill Balance");
				}
			}
			else
			{
				if(g_iClientTeam[client] == TEAM_T)
				{
					SwapPlayer(client, TEAM_CT, "Client Skill Balance");
				}
			}
		}
		++i;
	}
}
```

**Comment by NotJustin** [comment]
`int size = GetClientCountNoBots() / 2 - outliers;`
This line was wrong. Size is the number of clients that should be on each team, but this number does not exclude spectators (I do not put spectators on a team during a balance).

Replaced with:
`int size = (GetTeamClientCount(TEAM_T) + GetTeamClientCount(TEAM_CT)) / 2 - outliers;`

During this testing it also led me to fix another bug in how I was finding outliers. Basically, I was using the wrong client indices so I fixed it. Will close this issue if I go ~2 days without seeing this horrible misbalance happen.

However, I suspect something else. It could be either
`if (client != 0 && IsClientInGame(client) && !IsFakeClient(client) && (team = GetClientTeam(client)) != TEAM_SPEC && team != UNASSIGNED && !g_iOutlierClients[client])` this large conditional fails for clients that it should not fail for.

Or

`g_iClientTeam[client]`, which is the team that the client "belongs" on, is incorrect.
In the large code in the first comment, I use this value to check whether or not I should be swapping clients.
If this matches the team I'm trying to swap the client to, it means they already are on that team.
If that is the case, I don't swap them, but I do still increase the # of clients on that team.

If the issue is still occurring, I will replace this with `GetClientTeam` to see what team the client is truly on.

**Comment by NotJustin** [comment]
It looks like it is still occurring. I'm going to test if g_iClientTeam is correct like I mentioned above.

edit: Confirmed. In my logs I checked if g_iClientTeam is different from GetClientTeam when they're on T/CT, and it is. Now I need to figure out why it is different.

**Comment by NotJustin** [comment]
The teams are different because of the fix for spawn bug.

When a client joins a team, I need to know what team they joined. This is important. If the client joins spectate and then tries to join a team again, I must force them to join the team they were on before. Otherwise, people will try to stack good players on one team.

When a client fully connects to the server, I do `ClientCommand(client, "jointeam 0 %d, team")`. When I tried to update what team the client joined after this command, the client's team didn't change yet. I needed to add a delay. Pushing a fix for this.

**Comment by NotJustin** [comment]
Not fixed. Will keep debugging.

I'm just going to use GetClientTeam when swapping teams for now.

**Comment by NotJustin** [comment]
Okay, another thing about size was wrong and I have fixed it.

```
L 06/04/2020 - 16:06:50: [skillautobalance_debug.smx] Balance needed.
L 06/04/2020 - 16:06:50: [skillautobalance_debug.smx] Total clients: 15
L 06/04/2020 - 16:06:50: [skillautobalance_debug.smx] Size of T: 5
L 06/04/2020 - 16:06:50: [skillautobalance_debug.smx] Size of CT: 6
L 06/04/2020 - 16:06:50: [skillautobalance_debug.smx] Number of outliers: 1
L 06/04/2020 - 16:06:50: [skillautobalance_debug.smx] Approximate size of both teams minus outliers: 4
```

The size of each team should be 5, not 4. Total clients is 15, but number of players in game is only 5 + 6 = 11. 11 / 2 is 5.5 but integer drops the decimal, so it's 5. Then I was subtracting the outlier, so it became 5 - 1 = 4. The size of both teams was 4 when it should have been 5.

Before:
`int size = (GetTeamClientCount(TEAM_T) + GetTeamClientCount(TEAM_CT)) / 2 - outliers;`

After: 
`int size = (GetTeamClientCount(TEAM_T) + GetTeamClientCount(TEAM_CT) - outliers) / 2;`

Subtracting the outlier before dividing does this. 5 + 6 - 1 = 10 / 2 = 5.

I am still occasionally getting an error where g_iClientTeam is incorrect. I will explain what I want exactly:

I want to get what team the client is on right after I do Plugin_Continue in the command listener for "jointeam". Right now I'm putting a timer right before Plugin_Continue, but sometimes it still says the client is on UNASSIGNED (team index 0).

**Comment by NotJustin** [comment]
```
L 06/04/2020 - 17:50:30: [skillautobalance_debug.smx] Balance needed.
L 06/04/2020 - 17:50:30: [skillautobalance_debug.smx] Total clients: 12
L 06/04/2020 - 17:50:30: [skillautobalance_debug.smx] Size of T: 5
L 06/04/2020 - 17:50:30: [skillautobalance_debug.smx] Size of CT: 5
L 06/04/2020 - 17:50:30: [skillautobalance_debug.smx] Number of outliers: 1
L 06/04/2020 - 17:50:30: [skillautobalance_debug.smx] Approximate size of both teams minus outliers: 4
L 06/04/2020 - 17:50:30: [skillautobalance_debug.smx] Team sizes after: T: 6, CT: 4
```

12 clients, but 5 players on each team. So, only 10 players are in game.
Of the 10 players in game, 1 is an outlier. That leaves 9.
9/2 = 4.5, drop decimal, each team is size 4. When one team reaches 4 players, the rest of the players are dumped on the other team. It is supposed to exclude outliers though. But I have now found out that, I forgot to exclude outliers from the "dump remaining players" section. That is causing the issue. So now I will fix it.

I've said "this should fix the issue" for lots of things now but I am confident this time.

**Comment by NotJustin** [comment]
Did not mean to close this again.

No confirmation yet that this has fixed the issue, because the problem appears to have happened even with no outliers:

```
L 06/04/2020 - 15:25:25: [skillautobalance_debug.smx] Balance needed.
L 06/04/2020 - 15:25:25: [skillautobalance_debug.smx] Total clients: 16
L 06/04/2020 - 15:25:25: [skillautobalance_debug.smx] Size of T: 7
L 06/04/2020 - 15:25:25: [skillautobalance_debug.smx] Size of CT: 8
L 06/04/2020 - 15:25:25: [skillautobalance_debug.smx] Number of outliers: 0
L 06/04/2020 - 15:25:25: [skillautobalance_debug.smx] Approximate size of both teams minus outliers: 7
L 06/04/2020 - 15:25:25: [skillautobalance_debug.smx] Team sizes after: T: 6, CT: 9
```

**Comment by NotJustin** [comment]
Okay, I changed the while loop to look like this.

`while((totalSize % 2 == 0 && tCount < teamSize && ctCount < teamSize) || (totalSize % 2 == 1 && tCount <= teamSize && ctCount <= teamSize))`

totalSize is the # of players on T and CT combined, minus outliers.
teamSize is totalSize / 2

The left side of the or `||` is this:
`(totalSize % 2 == 0 && tCount < teamSize && ctCount < teamSize)`
total size is an even number, check the sizes in one way.

The right side of the or `||` is this:
`(totalSize % 2 == 1 && tCount <= teamSize && ctCount <= teamSize)`
total size is an odd number, check the sizes in different way.

It is hard for me to explain, but I think this is necessary. Testing now.

**Comment by NotJustin** [comment]
Still occurring. I got a 6v3 and a 6v8 today.

**Comment by NotJustin** [comment]
edit: I logged incorrectly. Will post better info in a few hours.

```
L 06/07/2020 - 22:16:05: [skillautobalance.smx] Balancing teams
L 06/07/2020 - 22:16:05: [skillautobalance.smx] Team sizes at start: T: 9, CT: 9
L 06/07/2020 - 22:16:05: [skillautobalance.smx] Team sizes before adding outliers: T: 11, CT: 7
L 06/07/2020 - 22:16:05: [skillautobalance.smx] Number of outliers: 0
L 06/07/2020 - 22:16:05: [skillautobalance.smx] Team sizes after: T: 11, CT: 7
```
Welp. There's some other flaw - it just isn't doing what I describe it to be doing.

**Comment by NotJustin** [comment]
Here's an update. I have narrowed down the issue, fixing a lot of the logic related to outliers. I will try to explain the current issue. As I am swapping players between teams, I keep track of the number of players I have swapped with a count.

One is `tCount` for number of players on terrorist.
One is `ctCount` for number of players on counter-terrorist.

The problem is, one player gets swapped twice. They are counted as being on both team. So, one team has one less player than I want, and the other team has one more player than I want.

I am stuck here. I have no idea why one player is being swapped twice. It looks like there is a duplicate index being created, but I never modify it myself after map start. The only time it is "modified" is when I am sorting the arrays, but those are built-in sourcemod functions.

Not sure what to do.
